### ISSUE TITLE
CSV ingest improvements

### ISSUE THREAD

**Issue by landreev** [IQSS/dataverse#3767]
(this issue is for a clearly defined, short term goal; not another generic "improve xxx" issue;) 

The original CSV parser was purposefully restrictive; strict formatting - one line per observation (no new lines in fields), fixed number of commas per line, etc. These requirements are no longer relevant. At the same time Gary specifically requested the CSV ingest to handle full text - with escaped new lines and rich punctuation, etc. Example: the file posts_all.tab (csv original) in  https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/QSZMPD.

One way to define the goal would be to say that any Google/Excel spreadsheet columns exported as CSV should be parseable by our ingest. (I will add more details on how they escape punctuation characters and such). 

A sensible way to achieve this would be to switch to some available open source parser (Apache seems like a good candidate), rather than maintaining our own.

**Comment by raprasad** [comment]
**code**

```python
>>> import pandas as pd
>>> df = pd.read_csv('posts_all.csv')
>>> df.columns
Index([u'What.is.the.city.', u'folder', u'file',
       u'What.is.the.name.of.the.organization.making.posts.', u'url',
       u'content', u'site', u'What.is.the.account.name.of.the.person.posting.',
       u'PostDate', u'category', u'textseg'],
      dtype='object')
>>> df['What.is.the.city.'].describe()
count         43757
unique            1
top       Zhanggong
freq          43757
Name: What.is.the.city., dtype: object
>>> df.describe()
         category
count  188.000000
mean     3.936170
std      0.381624
min      3.000000
25%      4.000000
50%      4.000000
75%      4.000000
max      5.000000
>>> df['What.is.the.name.of.the.organization.making.posts.'].describe()
count     43722
unique      247
top         网宣办
freq       9159
Name: What.is.the.name.of.the.organization.making.posts., dtype: object
```

etc, etc

**Comment by pdurbin** [comment]
@raprasad heh. @mercecrosas also just mentioned to us that we could use the `readxl` R package, which just reached 1.0: https://blog.rstudio.org/2017/04/19/readxl-1-0-0/ . Using either Python or R for this sounds more like #2331 to me.

Back at https://github.com/IQSS/dataverse/issues/585#issuecomment-291776481 @bencomp mentioned using https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVParser.html , which I believe is what @landreev was referring to above.

In short, we need to decide what's in scope for this issue. I thought when we estimated this issue as an 8 on Wednesday we were talking about the Java/Apache Commons route, not the Python or R route. Perhaps we should re-estimate this issue and/or estimate #2331.

**Comment by pdurbin** [comment]
@oscardssmith does your pull request at #3930 also fix #2626?

**Comment by oscardssmith** [comment]
No idea, I'll put it in as a test case. Is imagine it's fixed though

**Comment by landreev** [comment]
@oscardssmith the file I was talking about is the one mentioned in the initial description, at the top of the issues.

**Comment by pdurbin** [comment]
@oscardssmith does ingest of tab-separated values (tsv) files work with the code you added? See #1912.

**Comment by oscardssmith** [comment]
Not currently. In order to parse a csv, we have to assume a parsing for it. That said, I could theoretically add the ability to parse tsv's without too much code reuse. That said, that is definitely the bad type of scope creep right now.

**Comment by pdurbin** [comment]
@oscardssmith ok, once you put this issue in code review (for me this often means I'm sick of staring at the code and ready to move on to some other issue 😄 ), please leave a comment on #1912 suggesting an approach. Thanks.

**Comment by landreev** [comment]
The code in this PR is not immediately addressing #1912. But as I'm rewriting the documentation, I will mention that a way to ingest a TSV file is to convert it to CSV, and will explain how to do it cleanly. 

This issue was narrowly defined to improve the CSV parsing. As it's important and complex enough by itself. But once we are satisfied with how it's working, we could use it as a solution for TSV parser; by implementing a TSV reader that simply converts a tab-delimited file to CSV (while properly escaping all the commas first), transparently to the user, and then passes it to this reader. I'll mention it in #1912.

**Comment by landreev** [comment]
For the record, regarding my last comment, having looked at the new parser code some more, it *may* actually be fairly straightforward to make this ingest plugin handle both CSV and TSV... So we'll consider it, once ingesting CSV files is confirmed to be working as it should.

**Comment by oscardssmith** [comment]
It would be easy (especially with the latest rewrite). That said, we should merge this without the feature, and add it in a separate PR later if we decide we want it.

**Comment by landreev** [comment]
Agree (with the last comment); will open a separate issue - or, better yet, will use the existing issue 1912. 

But yes, it should be easy: the plugin can check the content type of the datafile it's reading; and pick the delimiter character to use based on that. 

```
if (DataFile.getContentType.equals("text/csv") || ...) {
   delimiterChar = ',';
} else if (DataFile.getContentType.equals("text/tab-separated-values" || ...) {
   delimiterChar ='\t'; 
} 
...
```

**Comment by landreev** [comment]
Please note that this issue is definitely not ready and may still need to go back into development. 
Putting it into review was still useful, if only to have another team member familiarize themselves with what's going on. (So who's actually reviewing it, @gdurand?) But as I said yesterday, since it deals with something as important as ingest, it cannot possibly go into production until tested against production data. And we weren't ready to start doing that until yesterday. I started a test batch job before I left yesterday, and it completed today (more on that below; but, long story short, there are inconsistencies and, possibly, bugs in the code). 

Aside from the code, unfortunately I didn't have time to update the CSV ingest documentation. Whether this is handled as part of this issue, or whether we open a new one, it must be addressed before the new ingest driver is released. Because our current doc still says we require super-strict formatting (no split lines; no escaped commas - same number of commas in every line.) As the absolute minimum, this line needs to be changed: 

All the lines that follow must contain the same number of comma-separated values as the first, variable name line.

And explain that, as of 4.7.1, we allow extra commas and extra newlines, as long as they are enclosed in double quotes. Depending on how fancy we want to be, we can add some examples and discussion. We should probably mention that we are now using the Apache CSV parser, and provide a link to their (Apache's) documentation, for those who are really interested in the technical nuances of their CSV parsing. (tagging @dlmurphy here, since you are probably going to be involved in this in some way)

Back to the testing:  

We appear to have about 11K ingested CSV files currently in production, about 100GB total. The query below was used to select the relevant information from the production database: 

```
SELECT d.id, p.authority, p.identifier, f.filesystemname, f.filesize, t.id, t.originalfileformat, t.unf from dvobject d, datafile f, dataset p, datatable t WHERE t.originalfileformat='text/csv'  AND d.id = f.id AND t.datafile_id = f.id AND d.owner_id = p.id
```

100GB is a bit bulky. So I sorted these 11K files by size, and selected the firsty 10,000. These resulted in 20GB of files - much more manageable, but still a very comprehensive sample. So I copied these onto dvn-vm5.hmdc, where I had the latest (as of yesterday) build from the csv ingest branch. (I will send the details of the logistics  - where the files live, how they are organized, where the batch scripts are, etc. separately, by email, as to not to mess up github with that stuff). 

In the ingest branch, I reactivated the test ingest API that I used for batch testing during the redevelopment of the stata and spss ingest plugins during for the initial 4.0 release. The API accepts the location of the file, runs the ingest code on it, and, *without saving the results in the database*, outputs the basic ingest info: 

```
# curl "http://localhost:8080/api/ingest/test/file?fileName=/nfs/dvn/testdata/csv/10.7910/DVN/0X97MO/15798e52843-42322b8e30fb.orig&fileType=text/csv"
NVARS: 2
NOBS: 2
UNF: UNF:6:XPpFYeXw6MSKgd+dvsjRZg==
VAR0 A character UNF:6:c0Rl7c6BtDUs3idtWabmug==
VAR1 1 numeric-discrete UNF:6:EDaMAZ1/p7Pk32kDe7b4PQ==
```

(it also saves the produced tabular file; next to the supplied location of the original; but with the ".tab" extension).
This allows to test the ingest code fairly quickly, without actually ingesting the files into the database. 

So, last night I ran a batch of 10,000 prod. CSV files,  starting with the smallest and progressing to the larger. The job completed in 16 hours, I believe. I'll start with the good news: Out of these 10K,  8,522 files produced identical tab files and identical UNFs. The bad news, however, is that the total of 971 did not ingest at all (failed to produced a tab file or variable metadata). And 507 files did ingest, having transformed the data into tab delimited format, but produced *different* UNFs (than what we have in production). 

It is still entirely possible that the new ingest driver is doing everything correctly. And that the inconsistencies are entirely due to something wrong with the old ingest... Case in point, the very first ingest failure was on this 2 column, 3 line micro-CSV file (datafile id 2970268 in the production db):

```
"","Random"
"Views",1249653
"Clicks",1495
"CTR",0.119633210179146
```

Looking at the first line, it must mean the old ingest was allowing empty variable names in the first line (as long as the empty space was enclosed in quotes??)... I feel it is safe to say, that this (the *old* behavior) is wrong. Empty variables should not be allowed...

Still, the remaining 1477 inconsistencies still need to be investigated as well. :( And something's telling me, that at least some of them are probably bugs in the new code. So, once again, whether this is addressed as part of this issue, or as a new one, it needs to be done. 

So, will send the details on the dvn-vm5 batch test setup, by email, to everybody involved.

**Comment by pdurbin** [comment]
> Please note that this issue is definitely not ready and may still need to go back into development

That's fine. I just moved it back to development. Thanks for the detailed comment @landreev as well as the internal "CSV ingest, status" email.

**Comment by dlmurphy** [comment]
I checked in with @oscardssmith and he filled me in on what the documentation might need. Thanks to his excellent counsel, I just added some updates that address the stuff @landreev mentioned in the comment above. I would definitely like to have someone besides Oscar or myself review the docs before they go live to ensure that we addressed everything that needs addressing.

**Comment by pdurbin** [comment]
@oscardssmith when you have a moment, can you please leave a comment at #3769 to give your opinion of what we do well and what we do poorly in terms of reporting ingest failures? Thanks!

**Comment by pdurbin** [comment]
It would be nice if the two files at http://dx.doi.org/10.7910/DVN/Y3WOOE  (2717466 and 2717467)  were explorable as tabular data. I just converted them from xlsx to csv using Excel for Mac version 15.13.3 (150815) and uploaded them to my laptop running branch `3767-CSV-injest-code` as of fd5bbf5 but unfortunately, both failed ingest for different reasons.

**OJS-Followup-2015.csv** failed with "Invalid variable names in the first line! - First line of a CSV file must contain a comma-separated list of the names of the variables." Here's screenshot:

![screen shot 2017-07-05 at 11 45 16 am](https://user-images.githubusercontent.com/21006/27873496-54d9f79c-617a-11e7-9cf8-1d26419eaf13.png)

**Publisher-Survey-Result-DEID.csv** failed with "Ingest succeeded, but failed to save the ingested tabular data in the database: Transaction aborted". Here's a screenshot followed by the exception in the Glassfish server.log file:

![screen shot 2017-07-05 at 11 45 29 am](https://user-images.githubusercontent.com/21006/27873472-454b6e0a-617a-11e7-9300-5a196aa7d088.png)


`Internal Exception: org.postgresql.util.PSQLException: ERROR: value too long for type character varying(255)`
`Error Code: 0`
`Call: INSERT INTO INGESTREPORT (ENDTIME, REPORT, STARTTIME, STATUS, TYPE, DATAFILE_ID) VALUES (?, ?, ?, ?, ?, ?)`
`        bind => [6 parameters bound]`

Also, I'm seeing some weird characters in OJS-Followup-2015.csv but @oscardssmith said it probably isn't affecting ingest:

![screen shot 2017-07-05 at 11 54 30 am](https://user-images.githubusercontent.com/21006/27873430-19b6e18e-617a-11e7-8ac2-0d5d495e84b8.png)

I'm uploading both files but appending ".txt" to the filename so that GitHub will accept them:

- [OJS-Followup-2015.csv.txt](https://github.com/IQSS/dataverse/files/1125287/OJS-Followup-2015.csv.txt)
- [Publisher-Survey-Result-DEID.csv.txt](https://github.com/IQSS/dataverse/files/1125289/Publisher-Survey-Result-DEID.csv.txt)

Incidentally, I did try uploading the original xlsx files as well but that failed ingest as well. This is out of scope for this issue (perhaps in scope for #585) instead, but I'll just mention that the error for both files was "Could not parse Excel/XLSX spreadsheet. Could not establish position index of a cell element unambiguously!".

**Comment by oscardssmith** [comment]
Here's a quick update on the pull request: I have gone through the many files Leonid tested against, and all of the ones that failed ingest would have failed before. Of the ones that succeeded, but produced different UNF's I have verified that the new one behaved correctly in 419 out of 507 cases. The other 88, I have been unable to verify which was correct, but they are all weird, and were likely ingested by hand anyway. Given that, I think this is ready to move on (although it makes sense to wait until @landreev is back before merging this)

**Comment by landreev** [comment]
Reviewing the results of the investigation - great job btw - I feel like we should consider at least one possible improvement/fix to be within the scope of this development iteration: All these cases of files that are NOT actually CSV, but tabular, that still get ingested as CSV files with just 1 column. The current code is doing it in a less broken way than what we were doing before... But it's still wrong to be doing this. 

So let's discuss this. As in whether we want to change the behavior now; and if we do - how? I'm assuming that all these ingests happen because the files get uploaded with file names that have the ".csv" extension. We then assume that it is indeed a CSV, and that we should give it a try as such - and accept the result unless the parser explicitly fails. It appears that we need to be more careful there - namely, if the parser works, but we only find 1 column, we should do some extra checks and see if that was a CSV file in the first place. Maybe, we should be counting the tabs as we parse? And, if we've reached the end of the file without finding any commas - maybe we should just try and parse again, using the tab as the delimiter character? (and what do we do if that works? ingest the file? Or reject it with "this is not a CSV file"?)

An alternative, of course, is to declare this out of scope, release as is, and open a new issue for this. As long as it is accounted for, and on the list of things that need to be addressed - because it really seems wrong, what we are doing now.

**Comment by oscardssmith** [comment]
I think it makes most sense to say out of scope, and get this pushed out the door.

**Comment by djbrooke** [comment]
If it's incrementally better, let's get it tested and released! :)

**Comment by landreev** [comment]
OK, I'll open a new issue for this; and finish the review of all the other cases, shortly.

**Comment by landreev** [comment]
Moving the issue into QA.

**Comment by kcondon** [comment]
OK, tested basic csv ingest and all tests passed:

1. Number, string, date ingest (YYYY-MM-DD)
2. Unrecognized date format treated as string
3. Missing values end up missing in resulting tab files except strings are empty quotes. On 2 Ravens, numeric missing show as invalid, dates and strings show as valid, consistent with prior version. Missing values are expressed in csv as no value without quotes.
4. Tested comma in string value by enclosing string in quotes
5. Tested carriage return in string value by enclosing in quotes, examining resultant tab file.
6. Tested accents in string.
7. Tested error handling of mistmatched columns and values.

So, in addition to the extensive automatic ingesting of production csv files this seems to pass testing.

One last minor request is to perhaps mention some of the above behavior in the docs, esp missing values, quotes to preserve commas and carriage returns, and support for UTF-8 characters.

**Comment by landreev** [comment]
OK, I'll quickly add the few things we've discussed with @dlmurphy to the guide; then will pass it to Derek for review.

**Comment by landreev** [comment]
@kcondon @dlmurphy : I wrote, and re-wrote a few paragraphs in the doc. @dlmurphy, please review for clarity, etc. Anything extra we want to document/discuss should definitely be addressed in the next dev. iteration. (there will be another iteration sometime soon, for adding direct support for tab-delimited files). As it is, I've already ended up writing more than I thought was in scope for this issue.
